### PR TITLE
ENG-11437NEW

### PR DESCRIPTION
**Pull request by Yusarin**
First commit, need more test

### CODE REVIEW

File: lib/python/voltcli/voltadmin.d/restore.py [hunk 1]
```diff
@@ -22,6 +22,12 @@
     options = (
         # Hidden option to restore the hashinator in addition to the tables.
         VOLT.BooleanOption('-S', '--hashinator', 'hashinator', None, default = False),
+        VOLT.StringListOption(None, '--tables', 'tables',
+                              'tables to include in the snapshot',
```
Comment (manjujames) [inline lib/python/voltcli/voltadmin.d/restore.py]: These are the tables to be restored, not to be included in the snapshot. The description is accurate for snapshot save, but not for restore. 
Same comment for skip tables.

File: lib/python/voltcli/voltadmin.d/restore.py [hunk 2]
```diff
@@ -30,11 +36,18 @@
...
+        return
     if runner.opts.hashinator:
         hashinator = 'true'
     else:
         hashinator = 'false'
     json_opts = ['{path:"%s",nonce:"%s",hashinator:"%s"}' % (runner.opts.directory, nonce, hashinator)]
+    if runner.opts.tables:
+        json_opts = ['{path:"%s",nonce:"%s",hashinator:"%s",tables:"%s"}' % (runner.opts.directory, nonce, hashinator, runner.opts.tables)]
+    elif runner.opts.skip_tables:
+        json_opts = ['{path:"%s",nonce:"%s",hashinator:"%s",skiptables:"%s"}' % (runner.opts.directory, nonce, hashinator, runner.opts.skip_tables)]
```
Comment (manjujames) [inline lib/python/voltcli/voltadmin.d/restore.py]: What happens if neither is specified?

Comment (Yusarin) [reply lib/python/voltcli/voltadmin.d/restore.py]: if neither specified, json_opts will be the default structure initialized at line 46.

File: lib/python/voltcli/voltadmin.d/restore.py (hunk 1 above)
Comment (Yusarin) [reply lib/python/voltcli/voltadmin.d/restore.py]: Will change the description soon.

File: lib/python/voltcli/voltadmin.d/restore.py (hunk 2 above)
Comment (manjujames) [reply lib/python/voltcli/voltadmin.d/restore.py]: Ah, I missed it. Thanks.

File: src/frontend/org/voltdb/sysprocs/SnapshotRestore.java [hunk 3]
```diff
@@ -1140,6 +1142,10 @@ private void handleUniqueViolations(String table_name,
         // Fetch all the savefile metadata from the cluster
         VoltTable[] savefile_data;
         savefile_data = performRestoreScanWork(path, pathType, nonce, dupsPath);
+        tablesStr = tablesStr.equals("") ? "" : tablesStr.substring(1, tablesStr.length() - 1);
```
Comment (manjujames) [inline src/frontend/org/voltdb/sysprocs/SnapshotRestore.java]: I think you can avoid a lot of this code and parsing using our own code if you use jsobj.optJSONArray above instead of jsobj.optString.

File: src/frontend/org/voltdb/sysprocs/SnapshotRestore.java [hunk 4]
```diff
@@ -2238,6 +2278,7 @@ private void verifyRestoreWorkResult(VoltTable[] results, VoltTable[] restore_re
                     for (int ii = 0; ii < restore_plan.length - 1; ii++) {
                         restore_plan[ii].siteId = actualToGenerated.get(restore_plan[ii].siteId);
                     }
+                    System.out.println("Performing restore for table: " + table.getTypeName());
```
Comment (manjujames) [inline src/frontend/org/voltdb/sysprocs/SnapshotRestore.java]: We need to take out System.out statements before merging.

File: src/frontend/org/voltdb/sysprocs/SnapshotRestore.java [hunk 5]
```diff
@@ -2056,8 +2062,31 @@ private final DigestScanResult performRestoreDigestScanWork(boolean isRecover)
...
 
     private Set<Table> getTablesToRestore(Set<String> savedTableNames,
-                                          StringBuilder commaSeparatedViewNamesToDisable) {
+                                          StringBuilder commaSeparatedViewNamesToDisable,
+                                          List<String> include,
+                                          List<String> exclude,
+                                          int filter) {
         Set<Table> tables_to_restore = new HashSet<Table>();
+
+        if(filter == 1) {
```
Comment (manjujames) [inline src/frontend/org/voltdb/sysprocs/SnapshotRestore.java]: I don't think you need this filter in addition to the include and exclude lists. You can just check if the include list has elements in it, instead of filter==1 and check if the exclude list has elements in it instead of filter==-1.

File: src/frontend/org/voltdb/sysprocs/SnapshotRestore.java [hunk 6]
```diff
@@ -2056,8 +2062,31 @@ private final DigestScanResult performRestoreDigestScanWork(boolean isRecover)
...
+            Set<String> newSet = new HashSet<>();
+            for(String s : include) {
+                if(savedTableNames.contains(s)) {
+                    newSet.add(s);
+                } else {
+                    SNAP_LOG.error("Table: " + s + " is not in the snapshot, skip this table.");
+                    System.err.println("ERROR: Table " + s + " is not in the snapshot, skip this table.");
```
Comment (manjujames) [inline src/frontend/org/voltdb/sysprocs/SnapshotRestore.java]: Should take out system.err statement. Also, you should throw VoltAbortException so that the restore will fail in this case.

File: src/frontend/org/voltdb/sysprocs/SnapshotRestore.java [hunk 7]
```diff
@@ -2056,8 +2062,31 @@ private final DigestScanResult performRestoreDigestScanWork(boolean isRecover)
...
+                }
+            }
+            savedTableNames = newSet;
+        } else if(filter == -1) {
+            for (String s : exclude) {
+                if(savedTableNames.contains(s)) {
+                    savedTableNames.remove(s);
+                }
```
Comment (manjujames) [inline src/frontend/org/voltdb/sysprocs/SnapshotRestore.java]: If the table to be excluded is not in the saved list, print an INFO level log message saying that the table was not found.

File: tests/frontend/org/voltdb/regressionsuites/TestSaveRestoreSysprocSuite.java [hunk 8]
```diff
@@ -1937,6 +1937,119 @@ public void testPartialSnapshotSave() throws Exception
         }
```
Comment (manjujames) [inline tests/frontend/org/voltdb/regressionsuites/TestSaveRestoreSysprocSuite.java]: This is a really big test. I think this is a good opportunity to start a TestRestore suite and put these new tests in there.

File: tests/frontend/org/voltdb/regressionsuites/TestSaveRestoreSysprocSuite.java [hunk 9]
```diff
@@ -1937,6 +1937,119 @@ public void testPartialSnapshotSave() throws Exception
...
+            fail("JSON exception" + e.getMessage());
+        }
+        VoltTable[] results;
+        results = client.callProcedure("@SnapshotRestore", jsObj.toString()).getResults();
+
+        while(results[0].advanceRow()) {
+            if (results[0].getString("RESULT").equals("FAILURE")) {
+                fail(results[0].getString("ERR_MSG"));
+            }
+            assert(!results[0].getString("TABLE").equals("REPLICATED_TESTER"));
```
Comment (manjujames) [inline tests/frontend/org/voltdb/regressionsuites/TestSaveRestoreSysprocSuite.java]: As a sanity test, could you also verify that restored table does contain partition_tester and it has 126 items in it?

File: tests/frontend/org/voltdb/regressionsuites/TestSaveRestoreSysprocSuite.java [hunk 10]
```diff
@@ -1937,6 +1937,119 @@ public void testPartialSnapshotSave() throws Exception
...
+        }
+
+        VoltTable[] results;
+        results = client.callProcedure("@SnapshotRestore", jsObj.toString()).getResults();
+
+        while(results[0].advanceRow()) {
+            if (results[0].getString("RESULT").equals("FAILURE")) {
+                fail(results[0].getString("ERR_MSG"));
+            }
+            assert(results[0].getString("TABLE").equals("REPLICATED_TESTER"));
```
Comment (manjujames) [inline tests/frontend/org/voltdb/regressionsuites/TestSaveRestoreSysprocSuite.java]: Also verify that partition_tester is not in here and verify that all 126 rows are in here.

File: tests/frontend/org/voltdb/regressionsuites/TestSaveRestoreSysprocSuite.java [hunk 11]
```diff
@@ -1937,6 +1937,119 @@ public void testPartialSnapshotSave() throws Exception
...
+
+        while(results[0].advanceRow()) {
+            if (results[0].getString("RESULT").equals("FAILURE")) {
+                fail(results[0].getString("ERR_MSG"));
+            }
+            assert(!results[0].getString("TABLE").equals("REPLICATED_TESTER"));
+        }
+    }
+
+    public void testSaveAndPartialRestoreTables() throws Exception {
```
Comment (manjujames) [inline tests/frontend/org/voltdb/regressionsuites/TestSaveRestoreSysprocSuite.java]: Name this test testSaveAndPartialRestoreTablesIncludeTables so that it is clear what this is testing.

**Comment by manjujames** [comment]
Did you manually test voltadmin changes? If not, could you please do that? 
Also make sure that you get the correct error message if you include both options.

File: tests/frontend/org/voltdb/regressionsuites/TestSaveRestoreSysprocSuite.java [hunk 12]
```diff
@@ -1937,6 +1937,119 @@ public void testPartialSnapshotSave() throws Exception
...
+        } catch (JSONException e) {
+            fail("JSON exception" + e.getMessage());
+        }
+
+        VoltTable[] results;
+        results = client.callProcedure("@SnapshotRestore", jsObj.toString()).getResults();
+
+        while(results[0].advanceRow()) {
+            if (results[0].getString("RESULT").equals("FAILURE")) {
+                fail(results[0].getString("ERR_MSG"));
```
Comment (manjujames) [inline tests/frontend/org/voltdb/regressionsuites/TestSaveRestoreSysprocSuite.java]: I added a comment to say that this case should throw VoltAbortException. So make sure that this will fail.

Also add a test where you specify non-existent tables as skiptables option.

File: tests/frontend/org/voltdb/regressionsuites/TestSaveRestoreSysprocSuite.java (hunk 8 above)
Comment (Yusarin) [reply tests/frontend/org/voltdb/regressionsuites/TestSaveRestoreSysprocSuite.java]: You mean I should create a new class extends TestSaveRestoreSysprocSuite, adding my tests here and skipping all other tests , just like TestSaveRestoreSysprocSuite in pro does ?

Comment (manjujames) [reply tests/frontend/org/voltdb/regressionsuites/TestSaveRestoreSysprocSuite.java]: No, you can start a TestRestoreSysprocSuite that just extends SaveRestoreBase.

Comment (Yusarin) [reply tests/frontend/org/voltdb/regressionsuites/TestSaveRestoreSysprocSuite.java]: My tests are using some methods defined in TestSaveRestoreSysprocSuite, so I should just copy them to my new suite ?

Comment (manjujames) [reply tests/frontend/org/voltdb/regressionsuites/TestSaveRestoreSysprocSuite.java]: We can talk about it in person, might be quicker that way.

File: src/frontend/org/voltdb/sysprocs/SnapshotRestore.java [hunk 13]
```diff
@@ -1374,7 +1380,20 @@ else if (!saveFileState.isConsistent())
             }
         }
 
-        results = performTableRestoreWork(savefile_state, ctx.getSiteTrackerForSnapshot(), isRecover);
+        try {
+            testValidIncludeTables(savefile_state, includeList);
```
Comment (manjujames) [inline src/frontend/org/voltdb/sysprocs/SnapshotRestore.java]: May be change this method name to validateIncludeTables.

File: src/frontend/org/voltdb/sysprocs/SnapshotRestore.java [hunk 14]
```diff
@@ -1374,7 +1380,20 @@ else if (!saveFileState.isConsistent())
...
+        } catch (VoltAbortException e) {
+            ColumnInfo[] result_columns = new ColumnInfo[2];
+            int ii = 0;
+            result_columns[ii++] = new ColumnInfo("RESULT", VoltType.STRING);
+            result_columns[ii++] = new ColumnInfo("ERR_MSG", VoltType.STRING);
+            results = new VoltTable[] { new VoltTable(result_columns) };
+            results[0].addRow("FAILURE", e.toString());
+            System.out.println(e.toString());
```
Comment (manjujames) [inline src/frontend/org/voltdb/sysprocs/SnapshotRestore.java]: Remove the System.out stmt.

File: src/frontend/org/voltdb/sysprocs/SnapshotRestore.java [hunk 15]
```diff
@@ -2056,8 +2075,27 @@ private final DigestScanResult performRestoreDigestScanWork(boolean isRecover)
...
+            for(String s : include) {
+                newSet.add(s);
+            }
+            savedTableNames = newSet;
+        } else if(exclude.size() > 0) {
+            for (String s : exclude) {
+                if(savedTableNames.contains(s)) {
+                    savedTableNames.remove(s);
+                } else {
+                    SNAP_LOG.info("Table: " + s + " does not match any savefile data");
```
Comment (manjujames) [inline src/frontend/org/voltdb/sysprocs/SnapshotRestore.java]: Change this message to "Table <name> does not exist in the saved snapshot"

**Comment by manjujames** [comment]
The changes look good to me other than the parsing code that looks unnecessary.

File: src/frontend/org/voltdb/sysprocs/SnapshotRestore.java [hunk 16]
```diff
@@ -1374,7 +1378,19 @@ else if (!saveFileState.isConsistent())
             }
         }
 
-        results = performTableRestoreWork(savefile_state, ctx.getSiteTrackerForSnapshot(), isRecover);
+        try {
+            ValidIncludeTables(savefile_state, includeList);
```
Comment (manjujames) [inline src/frontend/org/voltdb/sysprocs/SnapshotRestore.java]: It should be validateIncludeTables.

File: src/frontend/org/voltdb/sysprocs/SnapshotRestore.java [hunk 17]
```diff
@@ -2969,6 +3007,35 @@ public static ClientResponseImpl transformRestoreParamsToJSON(StoredProcedureInv
...
+            return ret;
+        }
+        for(int i = 0 ; i < raw.length() ; i++) {
+            String s = "";
+            try {
+                s = raw.getString(i).trim().toUpperCase();
+            } catch (JSONException e) {
+                e.printStackTrace();
+            }
+            if(s.length() > 0)
```
Comment (manjujames) [inline src/frontend/org/voltdb/sysprocs/SnapshotRestore.java]: Add braces here or make it one line.

File: src/frontend/org/voltdb/sysprocs/SnapshotRestore.java [hunk 18]
```diff
@@ -2969,6 +3007,35 @@ public static ClientResponseImpl transformRestoreParamsToJSON(StoredProcedureInv
...
+        List<String> ret = new ArrayList<>();
+        if(raw == null || raw.length() == 0) {
+            return ret;
+        }
+        for(int i = 0 ; i < raw.length() ; i++) {
+            String s = "";
+            try {
+                s = raw.getString(i).trim().toUpperCase();
+            } catch (JSONException e) {
+                e.printStackTrace();
```
Comment (manjujames) [inline src/frontend/org/voltdb/sysprocs/SnapshotRestore.java]: Don't do printStacktrace. This is an unexpected error, so you can throw RuntimeException (similar to how JSONException is handled in the beginning of the run method)

File: src/frontend/org/voltdb/sysprocs/SnapshotRestore.java [hunk 19]
```diff
@@ -2969,6 +3007,35 @@ public static ClientResponseImpl transformRestoreParamsToJSON(StoredProcedureInv
...
 
+    private List<String> tableOptParser(JSONArray raw) {
+        List<String> ret = new ArrayList<>();
+        if(raw == null || raw.length() == 0) {
+            return ret;
+        }
+        for(int i = 0 ; i < raw.length() ; i++) {
+            String s = "";
+            try {
+                s = raw.getString(i).trim().toUpperCase();
```
Comment (manjujames) [inline src/frontend/org/voltdb/sysprocs/SnapshotRestore.java]: You can move all statements with variable s inside this try catch, which will make it more readable and avoiding have to declare String s = ""

File: src/frontend/org/voltdb/sysprocs/SnapshotRestore.java [hunk 20]
```diff
@@ -200,7 +200,7 @@
 
     private final static String HASHINATOR_ALL_BAD = "All hashinator snapshots are bad (%s).";
     // These keep track of count per table that are reported restored by the snapshotrestore process.
-    static final Map<String, AtomicLong> m_reportStats = new HashMap<String, AtomicLong>();
+    public static final Map<String, AtomicLong> m_reportStats = new HashMap<String, AtomicLong>();
```
Comment (manjujames) [inline src/frontend/org/voltdb/sysprocs/SnapshotRestore.java]: Why did you change this to public?
